crates/feldera-storage/src/buffer_cache/cache.rs

```diff
@@ -4,24 +4,83 @@
 //! It forwards requests to a storage backend, but it also adds a buffer cache
 //! in front of the backend to serve reads faster if they are cached.
 
-use metrics::counter;
 use std::cell::RefCell;
 use std::collections::HashMap;
-use std::ops::Range;
-use std::rc::Rc;
-use wtinylfu::WTinyLfuCache;
+use std::env;
+use std::ops::{Deref, Range};
+use std::sync::Arc;
+use std::time::Instant;
+
+use metrics::{counter, histogram};
+use moka::future::Cache;
+use sysinfo::System;
 
 use crate::backend::{
     FileHandle, ImmutableFileHandle, StorageControl, StorageError, StorageRead, StorageWrite,
+    METRIC_BUFFER_CACHE_HIT, METRIC_BUFFER_CACHE_LATENCY, METRIC_BUFFER_CACHE_MISS,
+    METRIC_WRITES_FAILED,
 };
 use crate::buffer_cache::FBuf;
 
 /// The key for the cache is a tuple of (file_handle, offset, size), and
 /// identifies a slice in a file.
 type CacheKey = (i64, u64, usize);
 
-/// CachedBuffer (the values of the cache) are reference-counted buffers.
-type CachedBuffer = Rc<FBuf>;
+/// CachedFBuf (the values of the cache) are reference-counted buffers.
+type CachedFBuf = Arc<FBuf>;
+
+pub struct TinyLfuCache(Cache<CacheKey, CachedFBuf>);
+
+impl Deref for TinyLfuCache {
+    type Target = Cache<CacheKey, CachedFBuf>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Default for TinyLfuCache {
+    /// A default instance of the LruCache will check if the environment
+    /// variable `FELDERA_BUFFER_CACHE_BYTES` is set to determine the amount of
+    /// bytes to cache.
+    /// If not set, it will fall back to use approx. 1/3 of the system memory.
+    fn default() -> Self {
+        let bytes_from_env: Option<u64> = env::var_os("FELDERA_BUFFER_CACHE_BYTES")
+            .and_then(|v| v.into_string().ok().and_then(|s| s.parse::<u64>().ok()));
+        let cache_capacity = bytes_from_env.unwrap_or_else(|| {
+            let mut sys = System::new_all();
+            sys.refresh_all();
+            sys.total_memory() / 3
+        });
+
+        TinyLfuCache::with_capacity(cache_capacity as usize)
+    }
+}
+
+impl TinyLfuCache {
+    const BLOCK_SIZE: usize = 512;
+
+    /// Builds a TinyLFU cache with the given capacity.
+    ///
+    /// The capacity here is given in how many bytes of DRAM the sum of all
+    /// stored buffers should take up before we start to evict from it.
+    pub fn with_capacity(capacity: usize) -> Self {
+        TinyLfuCache(
+            Cache::builder()
+                .support_invalidation_closures()
```

Comment:
Reading https://docs.rs/moka/latest/moka/future/struct.CacheBuilder.html#method.support_invalidation_closures, I'm not sure we need this because I don't think we call [Cache::invalidate_entries_if](https://docs.rs/moka/latest/moka/future/struct.Cache.html#method.invalidate_entries_if)

Reply:
fixed